plugins/acl/dynamic-role-based/src/org/apache/cloudstack/acl/DynamicRoleBasedAPIAccessChecker.java

```diff
@@ -0,0 +1,167 @@
+
+    @Inject
+    private AccountService accountService;
+    @Inject
+    private RoleService roleService;
+
+    private List<PluggableService> services;
+    private Map<RoleType, Set<String>> annotationRoleBasedApisMap = new HashMap<>();
+
+    protected DynamicRoleBasedAPIAccessChecker() {
+        super();
+        for (RoleType roleType : RoleType.values()) {
+            annotationRoleBasedApisMap.put(roleType, new HashSet<String>());
+        }
+    }
+
+    private void denyApiAccess(final String commandName) throws PermissionDeniedException {
+        throw new PermissionDeniedException("The API does not exist or is blacklisted for the account's role. " +
+                "The account with is not allowed to request the api: " + commandName);
+    }
+
+    private boolean checkPermission(final List <? extends RolePermission> permissions, final RolePermission.Permission permissionToCheck, final String commandName) {
+        if (permissions == null || permissions.isEmpty() || Strings.isNullOrEmpty(commandName)) {
+            return false;
+        }
+        for (final RolePermission permission : permissions) {
+            if (permission.getPermission() != permissionToCheck) {
+                continue;
+            }
+            try {
+                if (permission.getRule().matches(commandName)) {
+                    return true;
+                }
+            } catch (InvalidParameterValueException e) {
+                LOGGER.warn("Invalid rule permission, please fix id=" + permission.getId() + " rule=" + permission.getRule());
+            }
+        }
+        return false;
+    }
+
+    public boolean isDisabled() {
+        return !roleService.isEnabled();
+    }
+
+    @Override
+    public boolean checkAccess(User user, String commandName) throws PermissionDeniedException {
+        if (isDisabled()) {
+            return true;
+        }
+        Account account = accountService.getAccount(user.getAccountId());
+        if (account == null) {
+            throw new PermissionDeniedException("The account id=" + user.getAccountId() + "for user id=" + user.getId() + "is null");
+        }
+
+        final Role accountRole = roleService.findRole(account.getRoleId());
+        if (accountRole == null || accountRole.getId() < 1L) {
+            denyApiAccess(commandName);
+        }
+
+        // Allow all APIs for root admins
+        if (accountRole.getRoleType() == RoleType.Admin && accountRole.getId() == RoleType.Admin.getId()) {
+            return true;
+        }
+
+        final List<RolePermission> rolePermissions = roleService.findAllPermissionsBy(accountRole.getId());
+
+        // Check for allow rules
+        if (checkPermission(rolePermissions, RolePermission.Permission.ALLOW, commandName)) {
+            return true;
+        }
+
+        // Check for deny rules
+        if (checkPermission(rolePermissions, RolePermission.Permission.DENY, commandName)) {
```

Comment:
@koushik-das (2) the `checkPermission` method is done in memory is fast enough for each of ALLOW and DENY permission type on the same list of role permissions